### ISSUE TITLE
CA1812 - Generates false positive when using a type as a type parameter

### ISSUE THREAD

**Issue by CoolDadTx** [dotnet/roslyn-analyzers#3210]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

[CA1812]

#### Repro steps

Code snippet
```csharp
public class SomeType
{
   public string GetValue ()
   {
      var data = SomeHelperMethodThatReturnsAGenericType<NestedType>();
      return data.Value;
   }

   private class NestedType
   {
      public string Value { get; set; }
   }   
}
```
1. Compile the code with CA1812 enabled.
2. Get a warning that `NestedType` is not used.
3. Clearly the type is used because it is passed as a type parameter. In this particular case it also is the return value so `data` is of that type.

#### Expected behavior

Types used as type parameters should not be considered to be unused by this rule. Additionally it seems odd that the analyzer didn't determine that `data` was that type and so it was also used as a local variable. 

#### Actual behavior

Get a warning of CA1812 with the recommendation of removing the type even though it is used as part of a generic type and a local variable is declared as that type as well.

**Comment by CoolDadTx** [comment]
Also seems to fire if a method returns it as part of a generic return type (e.g. `Task<T>`).

```csharp
internal class SomeType
{
   public Task<AnotherType> Foo () { … }
}

//Generates this warning
internal class AnotherType
{
}
```

**Comment by Evangelink** [comment]
I confirm that the latest master reproduces the issue mentioned. 

@mavasani Do we want to consider all types used in a generic type to be "instantiated"? Or do we want to try to be more specific and do some filtering to catch only `new T` and `default(T)`?

**Comment by CoolDadTx** [comment]
@Evangelink how would filtering work if the generic method you call is provided in a third party library? You wouldn't have any way of knowing that the method created an instance.

**Comment by Evangelink** [comment]
True but that's not a problem exclusive to generic types.

**Comment by CoolDadTx** [comment]
Yes but my point was that generic methods allow an internal (normally not visible to other assemblies) type to be accessible from another assembly. For non-generic methods you can easily limit the usage search to the current assembly and hence do thorough code analysis. But this limitation doesn't apply when generics are thrown into the mix. Nevertheless I'll let you all figure out the scenarios. But in the case I posted I was actually using a generic method defined in a separate assembly. I just want to make sure that the fix gets rid of the error in these situations.

**Comment by mavasani** [comment]
Just for context - this rule does not analyze projects that exposes its internals to other projects via InternalsVisibleToAttribute, so cross-assembly scenarios are not relevant here: https://github.com/dotnet/roslyn-analyzers/blob/f15404b312295e7cee16fb40b4c8d3f91f10f087/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs#L53-L65

**Comment by mavasani** [comment]
>  Do we want to consider all types used in a generic type to be "instantiated"? Or do we want to try to be more specific and do some filtering to catch only new T and default(T)?

@Evangelink I think we can bail out on types that are used as a type argument where corresponding type parameter has a `new` constraint.

**Comment by CoolDadTx** [comment]
@mavasani 

> so cross-assembly scenarios are not relevant here

But to clarify, generic methods allow cross assembly calls with internal types, hence why I mentioned that internal types used in other assemblies should not trigger the warning. Case in point, I have a helper library that takes JSON and returns a strongly typed object. The type itself is internal but the method resides in public class (extension actually) in a separate assembly. Hence the internal type will be instantiated by a separate assembly even though `InternalsVisibleTo` is not applied. 

Handling the `new` constraint would work in this case I believe but it will still generate false positives in other cases.

```csharp
//Assembly A
public static class Helper
{
   public static T DoWork<T> ( T value ) { return value; }
}

//Assembly B
void Foo ()
{
   var data = Helper.DoWork(new InternalType());
}

internal class InternalType { }
```

Currently, ignoring the generic type itself, the warning is generated even though `data` is of type `InternalType`. It seems like the warning shouldn't apply if the generic method has a `new` constraint OR it returns `T` and there is a definite assignment to something. Perhaps the analyzer is just getting confused about the return type right now and fixing the generic method behavior will resolve both issues. Nevertheless I would expect no warning in the above case. Thanks.

**Comment by mavasani** [comment]
@CoolDadTx I am not sure your last example will generate a false positive, as we will detect `new InternalType()` in assembly `B` during analysis.

For now, let us just address https://github.com/dotnet/roslyn-analyzers/issues/3210#issuecomment-575230885 and we can see if we have more cases that need to be handled.

**Comment by Andrew-Webb** [comment]
This one is driving me mad because VS 2019 isn't offering me any way to suppress this (false) message.

Can we at least have the ability to suppress?

The error is reported:-

![image](https://user-images.githubusercontent.com/3470891/85288860-5c975a00-b48e-11ea-88e9-ce469707dcaf.png)

The options I am given when I ctrl+dot on the wavy red line...

![image](https://user-images.githubusercontent.com/3470891/85288950-8c466200-b48e-11ea-92da-075a9f31db6f.png)

**Comment by CoolDadTx** [comment]
@Andrew-Webb not sure why it isn't showing up for you but you can suppress all these warnings using an attribute (preferred), pragma or suppression file.

```csharp
[System.Diagnostics.CodeAnalysis.SuppressMessage("Performance", "CA1812:Avoid uninstantiated internal classes", Justification = "<Pending>")]
internal class TestClass
{
}

#pragma warning disable CA1812
    internal class TestClass
    {
    }
#pragma warning restore CA1812
```

**Comment by mavasani** [comment]
Thanks @CoolDadTx - yes, you should manually apply a suppression with any of the above source suppression mechanisms that work for you.

@Andrew-Webb `CA1812` is a compilation level diagnostic - basically it needs analyzing the whole project before it can be reported. For compilation level diagnostics, we do not run these in live analysis while typing + do not show code fixes for them on Ctrl + Dot (including suppression fixes). The core reason is performance - we do not want to analyze the entire project every time you press Ctrl + Dot, that will block light bulb for a significant time and not be a good experience. You should treat `CA1812` as a build-only diagnostic which only gets reported and refreshes after doing an explicit build.

**Comment by Andrew-Webb** [comment]
Ok, thanks @CoolDadTx and @mavasani for the explanations.

I'm used to VS 2019 offering to suppress, so when it doesn't I get a bit lost.  But I get bitten by this warning quite a lot so I'm going to stash:-

> [System.Diagnostics.CodeAnalysis.SuppressMessage("Performance", "CA1812:Avoid uninstantiated internal classes", Justification = "<Pending>")]

in a handy place and reach for it when needed.

**Comment by Evangelink** [comment]
@mavasani I have re-read this thread and I am not entirely sure to understand if there is something to fix and if so what.

**Comment by Evangelink** [comment]
As far as I understand there are 3 cases discussed in this issue:

## No lightbulb option

> `CA1812` is a compilation level diagnostic - basically it needs analyzing the whole project before it can be reported. For compilation level diagnostics, we do not run these in live analysis while typing + do not compute code fixes for them on Ctrl + Dot (including suppression fixes). The core reason is performance - we do not want to analyze the entire project every time you press Ctrl + Dot, that will block light bulb for a significant time and not be a good experience. You should treat `CA1812` as a build-only diagnostic which only gets reported and refreshes after doing an explicit build.

I understand the performance issue but I don't think the experience is nice from a user point-of-view. I even forgot about this and couldn't understand what was going-on.

First of all I don't understand why we don't see at least the `Configure` part as it shouldn't have any performance impact.

Second, don't you think it would be possible to show the menu with the `Suppress` part allowing ONLY suppression of current issue?. This way you shouldn't have the performance penalty of having to process the full compilation. Or are you talking about the fact that we will have to recompile to be able to remove the issue after the suppression is applied? If so can't we simply hook like issue suppresions are doing? 

We could event think about having a different entry for `Document`, `Project` and `Assembly` where there would be an extra description warning the user that this might be a long running operation that will force the recompilation of the project. Personally, I wouldn't mind to have to wait if that's something I am intentionally doing. 

## Issue with generic types

Doing some tests with how the analyzer works for on v`5.0.1` gives the following result:

```csharp
public class SomeType
{
    public string GetValue()
    {
        var data1 = SomeHelperMethodThatReturnsAGenericTypeWithNewConstraint<NestedType1>();
        var data2 = SomeHelperMethodThatReturnsAGenericType<NestedType2>();
        return data1.Value + data2.Value;
    }

    private T SomeHelperMethodThatReturnsAGenericTypeWithNewConstraint<T>()
        where T : new()
    {
        throw new NotImplementedException();
    }

    private T SomeHelperMethodThatReturnsAGenericType<T>()
    {
        throw new NotImplementedException();
    }

    private class NestedType1 // No CA1812 raised here
    {
        public string Value { get; set; }
    }

    private class NestedType2 // CA1812 raised here
    {
        public string Value { get; set; }
    }
}
```

So it seems that the case with the `new` constraint is handled. Regarding the other case, I am a bit unsure what is the right way forward... In this case, it is sort of obvious we do use the `Value` property of the class `NestedType2` but I don't know if that would always be the case for a type used as generic type parameter.

@mavasani @sharwell How would you move forward from here?

**Comment by mavasani** [comment]
> No lightbulb option

This is a Roslyn feature request to allow diagnostics from last build, which cannot be produced from intellisense, to have associated code actions (at least the suppress and configure ones). This is non-trivial work, and given we have very few compilation end/build-only analyzers, it has always felt like there is not enough value for doing such a feature. Feel free to open a Roslyn issue if you still prefer this be taken to IDE design meeting.

**Comment by mavasani** [comment]
> Issue with generic types
>> So it seems that the case with the new constraint is handled

I think that is sufficient to close this issue.

**Comment by poke** [comment]
> So it seems that the case with the new constraint is handled.

I disagree. Generic type arguments are very often used in serialization settings but serializers will usually *not* use any type constraints mostly because they won’t rely on the generic typing to create the object.

For example, the following examples all cause the CA1812 warning:

```csharp
// Deserializing with System.Text.Json
JsonSerializer.Deserialize<InternalType>("…");

// Deserializing with Newtonsoft.Json
JsonConvert.DeserializeObject<InternalType>("…");

// Querying a database using Dapper
connection.Query<InternalType>("…");
```

I personally hit this with the third use case, since I was declaring an internal type that would only be used for a single query result with [Dapper](https://github.com/DapperLib/Dapper) which would then be post-processed into a different (public) type. I wouldn’t be surprised if the analyzer would also detect other situations, e.g. with [key-less entity types in EF Core](https://docs.microsoft.com/en-us/ef/core/modeling/keyless-entity-types?view=aspnetcore-5.0&tabs=data-annotations) (previously called query types) which are commonly used for read-only queries, e.g. against views or stored procedures.

**Comment by geeknoid** [comment]
On a related note, we're also getting warnings from something like this:

```csharp
    [TypeConverter(typeof(StringValueTypeConverter<PermissionLevel>))]
    public readonly struct PermissionLevel : IEquatable<PermissionLevel>
```

The warning claims that the internal type `StringValueTypeConverter<T>` is unused. Clearly it is.

**Comment by lonix1** [comment]
I hit this problem because a type was not instantiated by me, but rather by the DI container. Obviously the type is being used, its not "dead code".

Please reopen.

**Comment by lonix1** [comment]
Since this issue was closed even though the problem persists, I added a new issue #6086 for the same underlying problem, upvote that if you land here.
